go/store/datas/pull/pull_table_file_writer.go

```diff
@@ -0,0 +1,342 @@
+	AddTableFilesToManifest(ctx context.Context, fileIdToNumChunks map[string]int) error
+}
+
+type PullTableFileWriterStats struct {
+	// Bytes which are queued up to be sent to the destination but have not
+	// yet gone out on the wire.
+	BufferedSendBytes uint64
+
+	// Bytes which we sent to the destination. These have been delivered to
+	// the operating system to be sent to the destination database. This
+	// number never goes down. In the case that we have to retry an upload,
+	// for example, BufferedSendBytes will instead go up.
+	FinishedSendBytes uint64
+}
+
+func NewPullTableFileWriter(ctx context.Context, cfg PullTableFileWriterConfig) *PullTableFileWriter {
+	ret := &PullTableFileWriter{
+		cfg:        cfg,
+		addChunkCh: make(chan nbs.CompressedChunk),
+	}
+	ret.eg, ret.egCtx = errgroup.WithContext(ctx)
+	ret.eg.Go(ret.reqRespThread)
+	return ret
+}
+
+func (w *PullTableFileWriter) GetStats() PullTableFileWriterStats {
+	return PullTableFileWriterStats{
+		FinishedSendBytes: atomic.LoadUint64(&w.finishedSendBytes),
+		BufferedSendBytes: atomic.LoadUint64(&w.bufferedSendBytes),
+	}
+}
+
+// Adds the compressed chunk to the table files to be uploaded to the destination store.
+//
+// If there is a terminal error with uploading a table file, this method will
+// start returning a non-nil |error|.
+//
+// This method may block for arbitrary amounts of time if there is already a
+// lot of buffered table files and we are waiting for uploads to succeed before
+// creating more table files.
+func (w *PullTableFileWriter) AddCompressedChunk(ctx context.Context, chk nbs.CompressedChunk) error {
+	select {
+	case w.addChunkCh <- chk:
+		return nil
+	case <-ctx.Done():
+		return context.Cause(ctx)
+	case <-w.egCtx.Done():
+		return w.eg.Wait()
+	}
+}
+
+func (w *PullTableFileWriter) reqRespThread() (err error) {
+	reqCh := make(chan uploadReq)
+	respCh := make(chan tempTblFile)
+	manifestUpdates := make(map[string]int)
+
+	// Uploads that haven't been sent to a request thread yet.
+	var pendingUploads []uploadReq
+	// The count of uploads that have been accepted by request threads.
+	// Must reach 0 for uploads to be completed and the manifest update to
+	// proceed.
+	outstandingUploads := 0
+
+	// After the writer is closed, it stops accepting new chunks and waits
+	// for all in-flight uploads to complete.
+	closed := false
+
+	eg, ctx := errgroup.WithContext(w.egCtx)
+	for i := 0; i < w.cfg.ConcurrentUploads; i++ {
+		eg.Go(func() error {
+			return w.uploadThread(ctx, reqCh, respCh)
+		})
+	}
+
+	var curWr *nbs.CmpChunkTableWriter
+
+	defer func() {
+		close(reqCh)
+		egErr := eg.Wait()
+		if err == nil {
+			err = egErr
+		}
+
+		if curWr != nil {
+			// Cleanup dangling writer, whose contents will never be used.
+			curWr.Finish()
+			rd, _ := curWr.Reader()
+			if rd != nil {
+				rd.Close()
+			}
+		}
+	}()
+
+	for {
```

Comment:
The bulk of review time was trying to understand the state machine of this loop. Having channels going nil then coming back in certain situations make this for a pretty hard loop to follow.

I don't have anything to suggest other than add more comments to spare the next reader from needing to figure this out. Maybe break the `newChnk, ok := <-addChunkCh:` case into a seperate function? Having the select block be really clear about the different assumptions when you indifferent states may help.